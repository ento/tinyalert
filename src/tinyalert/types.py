import datetime
import enum
from functools import cached_property
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator, root_validator


class Point(BaseModel):
    metric_name: str
    time: datetime.datetime
    metric_value: Optional[float]
    absolute_max: Optional[float]
    absolute_min: Optional[float]
    relative_max: Optional[float]
    relative_min: Optional[float]
    measure_source: Optional[str]
    diffable_content: Optional[str]
    url: Optional[str]
    skipped: bool = False

    model_config = dict(from_attributes=True)


class SourceType(str, enum.Enum):
    exec_ = "exec"
    shell = "shell"
    file_ = "file"


class EvalType(str, enum.Enum):
    lines = "lines"
    raw = "raw"


class MeasureType(BaseModel):
    source_type: SourceType
    eval_type: EvalType

    @root_validator(pre=True)
    def parse_string_format(cls, v):
        attrs = v
        if isinstance(v, str):
            source_type, eval_type = v.split("-")
            attrs = dict(source_type=source_type, eval_type=eval_type)
        return attrs


class MetricConfig(BaseModel):
    name: str
    measure_source: str
    measure_type: MeasureType = Field(
        default_factory=lambda: MeasureType(
            source_type=SourceType.file_, eval_type=EvalType.lines
        )
    )
    diffable_source: str = None
    diffable_type: SourceType = SourceType.file_
    measure_source_is_diffable: bool = True
    absolute_max: Optional[float] = None
    absolute_min: Optional[float] = None
    relative_max: Optional[float] = None
    relative_min: Optional[float] = None
    url: Optional[str] = None


class Config(BaseModel):
    metrics: List[MetricConfig] = Field(default_factory=list)

    @field_validator("metrics", mode="before")
    def validate_metrics(cls, v):
        return [
            MetricConfig.model_validate(dict(name=name, **attrs))
            for name, attrs in v.items()
        ]


class MeasureResult(BaseModel):
    value: float
    source: str


class MetricDiff(BaseModel):
    metric_name: str
    diff: str


class ReportData(BaseModel):
    metric_name: str
    latest_value: Optional[float] = None
    previous_value: Optional[float] = None
    latest_values: List[float] = Field(default_factory=list)
    absolute_max: Optional[float] = None
    absolute_min: Optional[float] = None
    relative_max: Optional[float] = None
    relative_min: Optional[float] = None
    latest_diffable_content: Optional[str] = None
    previous_diffable_content: Optional[str] = None
    url: Optional[str] = None

    @cached_property
    def violates_absolute_max(self) -> bool:
        if self.latest_value is None:
            return False
        if self.absolute_max is None:
            return False
        return self.latest_value > self.absolute_max

    @cached_property
    def violates_absolute_min(self) -> bool:
        if self.latest_value is None:
            return False
        if self.absolute_min is None:
            return False
        return self.latest_value < self.absolute_min

    @cached_property
    def violates_relative_max(self) -> bool:
        latest_change = self.latest_change
        if latest_change is None:
            return False
        if self.relative_max is None:
            return False
        return latest_change > self.relative_max

    @cached_property
    def violates_relative_min(self) -> bool:
        latest_change = self.latest_change
        if latest_change is None:
            return False
        if self.relative_min is None:
            return False
        return latest_change < self.relative_min

    @cached_property
    def violates_absolute_limits(self) -> bool:
        return self.violates_absolute_max or self.violates_absolute_min

    @cached_property
    def violates_relative_limits(self) -> bool:
        return self.violates_relative_max or self.violates_relative_min

    @cached_property
    def violates_limits(self) -> bool:
        return self.violates_absolute_limits or self.violates_relative_limits

    @cached_property
    def latest_change(self) -> Optional[float]:
        if self.latest_value is None:
            return None
        if self.previous_value is None:
            return None
        return self.latest_value - self.previous_value
