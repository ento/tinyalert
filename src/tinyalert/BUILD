python_sources(
    run_goal_use_sandbox=False,
    dependencies=[":alembic_ini", "./alembic"],
)


resource(name="alembic_ini", source="alembic.ini")
