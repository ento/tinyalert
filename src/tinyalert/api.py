import datetime
import shlex
import subprocess
from pathlib import Path
from typing import Any, Dict, Iterator, List, Optional, Sequence, Tuple

from .db import DB
from .types import EvalType, MeasureResult, MeasureType, Point, ReportData, SourceType, GenerationMatchStatus


def push(
    db: DB,
    metric_name: str,
    value: Optional[float] = None,
    absolute_max: Optional[float] = None,
    absolute_min: Optional[float] = None,
    relative_max: Optional[float] = None,
    relative_min: Optional[float] = None,
    measure_source: Optional[str] = None,
    diffable_content: Optional[str] = None,
    url: Optional[str] = None,
    epoch: int = 0,
    generation: int = 0,
    tags: Dict[str, Any] = {},
) -> Point:
    p = Point(
        metric_name=metric_name,
        time=datetime.datetime.now(datetime.timezone.utc),
        metric_value=value,
        absolute_max=absolute_max,
        absolute_min=absolute_min,
        relative_max=relative_max,
        relative_min=relative_min,
        measure_source=measure_source,
        diffable_content=diffable_content,
        url=url,
        epoch=epoch,
        generation=generation,
        tags=tags,
    )
    return db.add(p)


def measure(source: str, method: MeasureType) -> MeasureResult:
    source_content = eval_source(source, method.source_type)
    if method.eval_type == EvalType.lines:
        return MeasureResult(
            value=len(source_content.splitlines()), source=source_content
        )
    if method.eval_type == EvalType.raw:
        return MeasureResult(value=float(source_content.strip()), source=source_content)
    raise Exception(f"Unknown measurement method: {method}")


def eval_source(source: str, method: SourceType) -> str:
    if method == SourceType.exec_:
        return subprocess.check_output(shlex.split(source), text=True).strip()
    if method == SourceType.shell:
        return subprocess.check_output(source, shell=True, text=True).strip()
    if method == SourceType.file_:
        return Path(source).read_text()
    raise Exception(f"Unknown source type: {method}")


def skip_latest(db: DB, metric_name: str):
    db.skip_latest(metric_name)


def combine(dest_db: DB, src_dbs: List[DB]):
    for src_db in src_dbs:
        for p in src_db.iter_all():
            dest_db.add(p)


def prune(db: DB, keep: int = 10) -> int:
    assert keep > 0, "keep must be greater than 0"
    return db.prune(keep=keep)


def recent(db: DB, count: int = 10) -> Iterator[Point]:
    assert count > 0, "count must be greater than 0"
    for p in db.recent(count=count):
        yield p


def gather_report_data(
    db: DB, metric_name: str, head_generation: Optional[int] = None
) -> ReportData:
    data = ReportData(metric_name=metric_name)
    points = list(db.recent(metric_name, count=None))

    if not points:
        return data

    data.latest_values = [p.metric_value for p in reversed(points[:10])]

    eligible_points = _iter_alert_eligible_points(points, head_generation)
    latest, generation_status = next(eligible_points, None)
    data.generation_status = generation_status
    if latest:
        data.latest_value = latest.metric_value
        data.absolute_max = latest.absolute_max
        data.absolute_min = latest.absolute_min
        data.relative_max = latest.relative_max
        data.relative_min = latest.relative_min
        data.latest_diffable_content = latest.diffable_content
        data.latest_url = latest.url
        data.latest_tags = latest.tags

    previous, _ = next(eligible_points, (None, None))
    if previous:
        data.previous_value = previous.metric_value
        data.previous_diffable_content = previous.diffable_content
        data.previous_url = previous.url
        data.previous_tags = previous.tags

    return data


def _iter_alert_eligible_points(
    all_points: Sequence[Point], head_generation: Optional[int] = None
) -> Iterator[Tuple[Point, GenerationMatchStatus]]:
    active_epoch = None
    generation_status = GenerationMatchStatus.NONE_SPECIFIED if head_generation is None else GenerationMatchStatus.NONE_MATCHED
    for i, p in enumerate(all_points):
        if i == 0:
            if head_generation is not None and p.generation == head_generation:
                generation_status = GenerationMatchStatus.MATCHED
            active_epoch = p.epoch
            yield p, generation_status
            continue
        if p.skipped:
            continue
        if p.epoch != active_epoch:
            break
        yield p, generation_status
