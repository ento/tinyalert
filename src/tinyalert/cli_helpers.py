from typing import Optional

import click
import pytimeparse2
from datetime import timedelta


class Duration(click.ParamType):
    name = "duration"

    def convert(self, value, param, ctx):
        if isinstance(value, timedelta):
            return value
        if isinstance(value, (float, int)):
            return timedelta(seconds=value)
        delta = self.from_string(value)
        if delta is None:
            self.fail(f"Invalid duration: {value!r}", param, ctx)
        else:
            return delta

    @staticmethod
    def from_string(value: str) -> Optional[timedelta]:
        secs = pytimeparse2.parse(value)
        return timedelta(seconds=secs) if secs is not None else None
