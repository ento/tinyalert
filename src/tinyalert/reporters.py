import difflib
import textwrap

from tabulate import tabulate
from termcolor import colored

from .types import GenerationMatchStatus, MetricDiff, ReportData


class TableReporter:
    header = dict(
        status_character="",
        metric_name="Name",
        latest_value="Value",
        abs_limits="Thresholds",
        change="Change",
        rel_limits="Thresholds",
        details="Details",
    )

    def __init__(self):
        self.rows = []

    def add(self, report_data: ReportData) -> None:
        self.rows.append(self._make_row(report_data))

    def _make_row(self, report_data: ReportData) -> dict:
        row = dict(
            status_character=report_data.status_character,
            metric_name=report_data.metric_name,
            latest_value="-",
            abs_limits="-",
            change="-",
            rel_limits="-",
            details="",
        )

        if report_data.latest_value is not None:
            value_string = f"{report_data.latest_value:g}"
            colored_value_string = (
                (colored(value_string, "red") + " [!]")
                if report_data.violates_absolute_limits
                else value_string
            )
            row["latest_value"] = colored_value_string

        if (
            report_data.previous_value is not None
            and report_data.generation_status != GenerationMatchStatus.NONE_MATCHED
        ):
            change_string = f"{report_data.latest_change:+g}"
            colored_change_string = (
                (colored(change_string, "yellow") + " [!]")
                if report_data.violates_relative_limits
                else change_string
            )
            row["change"] = colored_change_string

        if (report_data.absolute_min is not None) or (
            report_data.absolute_max is not None
        ):
            abs_thresholds = []
            if report_data.absolute_min is not None:
                abs_thresholds.append(f"{report_data.absolute_min}<=v")
            if report_data.absolute_max is not None:
                abs_thresholds.append(f"v<={report_data.absolute_max}")
            row["abs_limits"] = ", ".join(abs_thresholds)

        if (report_data.relative_min is not None) or (
            report_data.relative_max is not None
        ):
            rel_thresholds = []
            if report_data.relative_min is not None:
                rel_thresholds.append(f"{report_data.relative_min}<=Δ")
            if report_data.relative_max is not None:
                rel_thresholds.append(f"Δ<={report_data.relative_max}")
            row["rel_limits"] = ", ".join(rel_thresholds)

        details = []
        if (
            report_data.previous_url
            and report_data.generation_status != GenerationMatchStatus.NONE_MATCHED
        ):
            details.append(f"[baseline]({report_data.previous_url})")
        if report_data.latest_url:
            title = (
                "baseline"
                if report_data.generation_status == GenerationMatchStatus.NONE_MATCHED
                else "latest"
            )
            details.append(f"[{title}]({report_data.latest_url})")
        row["details"] = " \| ".join(details)

        return row

    def get_value(self) -> str:
        header = dict(self.header)
        rows = [dict(row) for row in self.rows]
        if not any(row["details"] for row in rows):
            del header["details"]
            for row in rows:
                del row["details"]
        return tabulate(rows, header, tablefmt="github")


class ListReporter:
    def __init__(self):
        self.items = []

    def add(self, report_data: ReportData) -> None:
        if report_data.violates_absolute_max:
            self.items.append(
                (
                    f"{report_data.metric_name} latest value {report_data.latest_value:g}:"
                    f" violates absolute max value of {report_data.absolute_max:g}"
                )
            )

        if report_data.violates_absolute_min:
            self.items.append(
                (
                    f"{report_data.metric_name} latest value {report_data.latest_value:g}:"
                    f" violates absolute min value of {report_data.absolute_min:g}"
                )
            )

        if report_data.violates_relative_max:
            self.items.append(
                (
                    f"{report_data.metric_name} changed by {report_data.latest_change:+g}:"
                    f" violates relative max value of {report_data.relative_max:+g}"
                )
            )

        if report_data.violates_relative_min:
            self.items.append(
                (
                    f"{report_data.metric_name} changed by {report_data.latest_change:+g}:"
                    f" violates relative min value of {report_data.relative_min:+g}"
                )
            )

    def get_value(self) -> str:
        return "\n".join([f"- {message}" for message in self.items])


class DiffReporter:
    def __init__(self):
        self.diffs = []

    def add(self, report_data: ReportData) -> None:
        if not report_data.violates_limits:
            return
        latest_lines = (
            report_data.latest_diffable_content.splitlines(keepends=False)
            if report_data.latest_diffable_content is not None
            else []
        )
        previous_lines = (
            report_data.previous_diffable_content.splitlines(keepends=False)
            if report_data.previous_diffable_content is not None
            else []
        )
        diff = difflib.unified_diff(
            previous_lines, latest_lines, "previous", "latest", lineterm=""
        )
        self.diffs.append(
            MetricDiff(metric_name=report_data.metric_name, diff="\n".join(diff))
        )

    def get_value(self) -> str:
        return "\n".join(
            [self._format_metric_diff(diff) for diff in self.diffs if diff.diff]
        )

    def _format_metric_diff(self, metric_diff: MetricDiff) -> str:
        return textwrap.dedent(
            """
        <details><summary>{metric_name} diff</summary>

        ```diff
        {diff}
        ```

        </details>
        """
        ).format(**metric_diff.model_dump())


class StatusReporter:
    def __init__(self):
        self.success = True

    def add(self, report_data: ReportData) -> None:
        if report_data.violates_limits:
            self.success = False

    def get_value(self) -> bool:
        return self.success
