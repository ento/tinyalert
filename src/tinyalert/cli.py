import sys
import json
from pathlib import Path
from typing import List, Optional

import click
import tomli

from . import api, db
from .reporters import DiffReporter, TableReporter, ListReporter, StatusReporter
from .types import Config, IgnoreType, ReportData


@click.group()
@click.argument("db_path", type=click.Path(exists=False), envvar="TINYALERT_DB")
@click.pass_context
def cli(ctx, db_path):
    ctx.obj = db.DB(db_path)
    ctx.obj.migrate()


@cli.command()
@click.argument("metric_name")
@click.option("--value", default=None, type=float, help="Measurement value")
@click.option("--abs-max", default=None, type=float, help="Absolute max (inclusive)")
@click.option("--abs-min", default=None, type=float, help="Absolute min (inclusive)")
@click.option("--rel-max", default=None, type=float, help="Relative max (inclusive)")
@click.option("--rel-min", default=None, type=float, help="Relative min (inclusive)")
@click.option("--source", default=None, help="Raw content of the measured thing")
@click.option(
    "--diffable", default=None, help="Text content suitable for showing diffs"
)
@click.option("--url", default=None, help="URL")
@click.option(
    "--ignore",
    default=None,
    type=click.Choice(IgnoreType),
    help="Do not alert on this data point",
)
@click.pass_context
def push(
    ctx,
    metric_name,
    value,
    abs_max,
    abs_min,
    rel_max,
    rel_min,
    source,
    diffable,
    url,
    ignore,
):
    if value is None:
        value = float(sys.stdin.read().strip())

    if not value:
        click.echo("No value provided")
        ctx.exit(1)

    api.push(
        db=ctx.obj,
        metric_name=metric_name,
        value=value,
        absolute_max=abs_max,
        absolute_min=abs_min,
        relative_max=rel_max,
        relative_min=rel_min,
        ignore=ignore,
        measure_source=source,
        diffable_content=diffable,
        url=url,
    )


def split_values(ctx, param, value):
    if not value:
        return None
    return [v.strip() for v in value.split(",")]


@cli.command()
@click.option(
    "--config", "-c", "config_path", type=click.Path(exists=True, path_type=Path)
)
@click.option(
    "--metrics",
    default=None,
    callback=split_values,
    help="Comma-separated names of metrics to measure",
)
@click.option("--url", default=None, help="URL")
@click.option(
    "--ignore",
    type=click.Choice(IgnoreType),
    default=None,
    help="Do not alert on this data point",
)
@click.pass_context
def measure(
    ctx: click.Context,
    config_path: Path,
    metrics: Optional[List[str]],
    url: Optional[str],
    ignore: Optional[IgnoreType],
):
    raw = tomli.loads(Path(config_path).read_text())
    config = Config.model_validate(raw)
    metric_configs_by_name = {metric.name: metric for metric in config.metrics}
    if metrics and set(metrics) - set(metric_configs_by_name.keys()):
        click.echo(
            "Unknown metrics: {}".format(
                ", ".join(set(metrics) - set(metric_configs_by_name.keys()))
            ),
            err=True,
        )
        ctx.exit(1)
    metrics_to_measure = metric_configs_by_name.keys() if metrics is None else metrics
    for metric_name in metrics_to_measure:
        metric = metric_configs_by_name[metric_name]
        result = api.measure(metric.measure_source, metric.measure_type)
        diffable_content = None
        if metric.diffable_source:
            diffable_content = api.eval_source(
                metric.diffable_source, metric.diffable_type
            )
        api.push(
            ctx.obj,
            metric.name,
            value=result.value,
            absolute_max=metric.absolute_max,
            absolute_min=metric.absolute_min,
            relative_max=metric.relative_max,
            relative_min=metric.relative_min,
            ignore=ignore,
            measure_source=result.source,
            diffable_content=diffable_content,
            url=url,
        )


@cli.command()
@click.argument("src_db_path", nargs=-1)
@click.pass_context
def combine(ctx, src_db_path):
    src_dbs = [db.DB(p) for p in src_db_path]
    api.combine(ctx.obj, src_dbs)


@cli.command()
@click.option("--json", "output_format", flag_value="json")
@click.pass_context
def recent(ctx, output_format):
    for p in api.recent(ctx.obj):
        if output_format == "json":
            print(p.model_dump_json())
        else:
            print("{time} {metric_name} {metric_value}".format(**p.dict()))


@cli.command()
@click.option("--format", "output_format", default="json")
@click.pass_context
def report(ctx, output_format):
    list_reporter = ListReporter()
    table_reporter = TableReporter()
    diff_reporter = DiffReporter()
    status_reporter = StatusReporter()

    for metric_name in ctx.obj.iter_metric_names():
        report_data = api.gather_report_data(ctx.obj, metric_name)
        table_reporter.add(report_data)
        list_reporter.add(report_data)
        diff_reporter.add(report_data)
        status_reporter.add(report_data)

    print(table_reporter.get_value())
    print(diff_reporter.get_value())
    print(list_reporter.get_value())
    if not status_reporter.get_value():
        ctx.exit(1)


@cli.command()
@click.option("--keep", type=int, required=True, help="How many points to keep")
@click.pass_context
def prune(ctx, keep):
    count = api.prune(ctx.obj, keep=keep)
    click.echo(f"Pruned {count} points in total", err=True)


@cli.command(
    context_settings=dict(
        ignore_unknown_options=True,
    )
)
@click.option("--help", is_flag=True)
@click.argument("alembic_args", nargs=-1, type=click.UNPROCESSED)
@click.pass_context
def migrate(ctx, alembic_args, help):
    args = list(alembic_args) + (["--help"] if help else [])
    ctx.obj.run_alembic(*args)
