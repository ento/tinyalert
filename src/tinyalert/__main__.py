from .cli import cli


def main():
    cli(auto_envvar_prefix="TINYALERT")


if __name__ == "__main__":
    main()
