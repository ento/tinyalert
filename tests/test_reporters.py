from unittest.mock import MagicMock

import pytest

from tinyalert.reporters import TableReporter, ListReporter, DiffReporter, StatusReporter
from tinyalert.types import ReportData


def test_table_reporter_outputs(snapshot):
    reporter = TableReporter()
    reporter.add(ReportData(metric_name="null"))
    reporter.add(ReportData(
        metric_name="kitchen-sink",
        latest_value=2,
        previous_value=1,
        latest_values=[0,1,2],
        absolute_max=0,
        absolute_min=3,
        relative_max=0,
        relative_min=2,
        url="http://example.com/",
    ))

    assert reporter.get_value() == snapshot


def test_list_reporter_outputs(snapshot):
    reporter = ListReporter()
    reporter.add(ReportData(metric_name="null"))
    reporter.add(ReportData(
        metric_name="kitchen-sink",
        latest_value=2,
        previous_value=1,
        latest_values=[0,1,2],
        absolute_max=0,
        absolute_min=3,
        relative_max=0,
        relative_min=2,
        url="http://example.com/",
    ))

    assert reporter.get_value() == snapshot


def test_diff_reporter_outputs(snapshot):
    reporter = DiffReporter()
    reporter.add(ReportData(metric_name="null"))
    reporter.add(ReportData(
        metric_name="kitchen-sink",
        latest_value=2,
        absolute_max=0,
        latest_diffable_content="aaa",
        previous_diffable_content="bbb",
    ))

    assert reporter.get_value() == snapshot


@pytest.mark.parametrize("violations,expected", [
    ([], True),
    ([False], True),
    ([True], False),
    ([True, False], False),
    ([False, True], False),
    ([False, False], True),
    ([True, True], False),
])
def test_status_reporter_outputs(violations, expected):
    reporter = StatusReporter()
    for violation in violations:
        reporter.add(MagicMock(violates_limits=violation))

    assert reporter.get_value() == expected
