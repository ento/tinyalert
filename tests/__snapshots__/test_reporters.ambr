# serializer version: 1
# name: test_diff_reporter_outputs
  '''
  
  <details><summary>kitchen-sink diff</summary>
  
  ```diff
  --- previous
  +++ latest
  @@ -1 +1 @@
  -bbb
  +aaa
  ```
  
  </details>
  
  '''
# ---
# name: test_diff_reporter_outputs_nothing_when_no_diffable_content
  ''
# ---
# name: test_list_reporter_outputs
  '''
  - kitchen-sink latest value 2: violates absolute max value of 0
  - kitchen-sink latest value 2: violates absolute min value of 3
  - kitchen-sink changed by +1: violates relative max value of +0
  - kitchen-sink changed by +1: violates relative min value of +2
  '''
# ---
# name: test_table_reporter_omits_details_column_when_possible
  '''
  |    | Name   | Value   | Thresholds   | Change   | Thresholds   | Trend   |
  |----|--------|---------|--------------|----------|--------------|---------|
  | -  | null   | -       | -            | -        | -            |         |
  '''
# ---
# name: test_table_reporter_outputs
  '''
  |    | Name         | Value   | Thresholds     | Change   | Thresholds     | Trend   | Details                                                                    |
  |----|--------------|---------|----------------|----------|----------------|---------|----------------------------------------------------------------------------|
  | -  | null         | -       | -              | -        | -              |         |                                                                            |
  | ▵  | kitchen-sink | 2 [!]   | 3.0<=v, v<=0.0 | +1 [!]   | 2.0<=Δ, Δ<=0.0 | ▁▄█     | [baseline](http://example.com/prev) \| [latest](http://example.com/latest) |
  '''
# ---
# name: test_table_reporter_outputs_latest_value_when_generation_didnt_match
  '''
  |    | Name         |   Value | Thresholds     | Change   | Thresholds     | Trend   | Details                               |
  |----|--------------|---------|----------------|----------|----------------|---------|---------------------------------------|
  | -  | kitchen-sink |       2 | 3.0<=v, v<=0.0 | -        | 2.0<=Δ, Δ<=0.0 | ▁▄█     | [baseline](http://example.com/latest) |
  '''
# ---
